### PR TITLE
[rostime] add SteadyTime

### PR DESCRIPTION
**Pull request by flixr**
Add a new SteadyTime wich uses a steady/monotonic clock to prevent issues with time jumps.

Replacement for #55 with MonotonicTime renamed to SteadyTime and added Windows implementation.

### CODE REVIEW

**Comment by flixr** [comment]
I didn't want to use the C++11 features here as I'm not sure if we can/should use them for roscpp yet.
And I'm hoping this will also applied to indigo (we are currently stuck on Ubuntu 14.04 on the Nvidia Jetson, as there are no cuda drivers for newer Ubuntu versions), where a C++11 compatible compiler is not yet required.

File: rostime/include/ros/time.h [hunk 1]
```diff
@@ -256,8 +256,43 @@ namespace ros
...
+   *
+   * Not affected by ROS time.
+   *
+   * ros::TimeBase provides most of its functionality.
+   */
+  class ROSTIME_DECL SteadyTime : public TimeBase<SteadyTime, WallDuration>
+  {
+    public:
+      SteadyTime()
+              : TimeBase<SteadyTime, WallDuration>()
```
Comment (dirk-thomas) [inline rostime/include/ros/time.h]: Please use two space indentation to be consistent with the existing code.

File: rostime/src/time.cpp [hunk 2]
```diff
@@ -93,8 +93,8 @@ namespace ros
    * These have only internal linkage to this translation unit.
    * (i.e. not exposed to users of the time classes)
    */
-  void ros_walltime(uint32_t& sec, uint32_t& nsec) 
-#ifndef WIN32    
+  void ros_walltime(uint32_t& sec, uint32_t& nsec)
```
Comment (dirk-thomas) [inline rostime/src/time.cpp]: Please avoid any unrelated whitespace changes in your PR.

**Review by dirk-thomas — COMMENTED** [review]
Please add unit tests covering the new functionality.

**Comment by flixr** [comment]
Fixed the whitespace stuff... will look into creating unit tests, although I have no idea on how to check if time is really monotonic, since I don't think we can reset the system time in the tests...

**Comment by flixr** [comment]
Anything else to do here?

**Comment by flixr** [comment]
Is it OK to use WallDuration with SteadyTime or should we add a SteadyDuration as well? There would be no difference between WallDuration and SteadyDuration though...

**Comment by dirk-thomas** [comment]
I think most importantly this needs to be tested in context of the full stack up to `bond_core` to make sure it satisfies the desired use case. @mikaelarguedas FYI

**Comment by mikaelarguedas** [comment]
I tested the changes to use SteadyTimer all the way up to bond_core. It works as expected (bonds don't break on forward time jumps, but if you jump back in time the bond will stay alive until the deadline is reached). I did notice a few unexpected nodelet crashes when using this so I'll do some more testing on Monday before +1 ing this.

**Comment by mikaelarguedas** [comment]
Update:
Looks like there are still situations where time jumps cause this to fail: 
```
[DEBUG] [1490639010.653929481]: SteadyTimer deregistering callbacks.
[DEBUG] [1490643600.000426555]: Time jumped forward by [25.143854] for timer of period [1.000000], resetting timer (current=1524518.175054, next_expected=1524493.031200)
[DEBUG] [1490643600.000492923]: Time jumped forward by [25.042610] for timer of period [1.000000], resetting timer (current=1524518.175106, next_expected=1524493.132497)
[ INFO] [1490643600.087787003]: Bond broken, exiting
```
It shows up only with nodelets but not with simple processes tied together with a bond. Needs more investigation.

Can be reproduced by running:
```
roslaunch nodelet_tutorial_math plus.launch
```
and jumping system time forward:
```
sudo date --set="2017-03-28 14:00:00.000"
```

**Comment by flixr** [comment]
@mikaelarguedas thanks a lot for testing!

The SteadyTime class of this PR should be ok, but I added a fix for the SteadyTimer in https://github.com/ros/ros_comm/pull/1014

**Comment by flixr** [comment]
@mikaelarguedas any updates here?
We have these 3 patchsets running here on all our devices and haven't encountered any problems so far.

**Comment by mikaelarguedas** [comment]
@flixr Sorry didn't have time to do another round of testing on this. Did you perform tests with nodelets as well or only on simple bond connection ? (IIRC I faced problem only with nodelets and multiple forward time jumps)

I'll try to test this today or tomorrow and will report here

**Comment by flixr** [comment]
I tested the SteadyTimer callbacks directly and nodelets...

**Comment by mikaelarguedas** [comment]
I tested the same here (cb, bonds only and nodelets) and couldn't make if fail. :clap: :tada: thanks for pounding on this that's a great improvement! :rocket:
@dirk-thomas What is the porting strategy you consider here ? Do you consider back-porting this to all active distros ?

File: rostime/src/time.cpp [hunk 3]
```diff
@@ -167,6 +167,40 @@ namespace ros
     nsec = nsec_sum;
 #endif
   }
+
+  void ros_steadytime(uint32_t& sec, uint32_t& nsec)
+#ifndef WIN32
```
Comment (dirk-thomas) [inline rostime/src/time.cpp:172]: This condition seems to be inverted. The exception is thrown in the Windows block.

File: rostime/src/time.cpp [hunk 4]
```diff
@@ -167,6 +167,40 @@ namespace ros
...
+    throw(NoHighPerformanceTimersException)
+#endif
+  {
+#ifndef WIN32
+    timespec start;
+    clock_gettime(CLOCK_MONOTONIC, &start);
+    sec  = start.tv_sec;
+    nsec = start.tv_nsec;
+#else
+    static LARGE_INTEGER cpu_frequency, performance_count;
```
Comment (dirk-thomas) [inline rostime/src/time.cpp:182]: Why should these variables be static? A local variable would avoid problems of the static variable being used by multiple concurrent callers.

File: rostime/src/time.cpp [hunk 5]
```diff
@@ -167,6 +167,40 @@ namespace ros
...
+    // These should not ever fail since XP is already end of life:
+    // From https://msdn.microsoft.com/en-us/library/windows/desktop/ms644905(v=vs.85).aspx and
+    //      https://msdn.microsoft.com/en-us/library/windows/desktop/ms644904(v=vs.85).aspx:
+    // "On systems that run Windows XP or later, the function will always succeed and will
+    //  thus never return zero."
+    QueryPerformanceFrequency(&cpu_frequency);
```
Comment (dirk-thomas) [inline rostime/src/time.cpp:188]: The return value of the function should be checked.

File: rostime/src/time.cpp [hunk 6]
```diff
@@ -167,6 +167,40 @@ namespace ros
...
+    if (cpu_frequency.QuadPart == 0) {
+      throw NoHighPerformanceTimersException();
+    }
+    QueryPerformanceCounter(&performance_count);
```
Comment (dirk-thomas) [inline rostime/src/time.cpp:192]: The return value of the function should be checked.

**Comment by dirk-thomas** [comment]
The change in this repo is easier since it only adds API. The PRs in other repos building on top of this are more challenging since they change behavior. Therefore I plan to roll them out distro by distro. So first Lunar and Kinetic, waiting for PRs on top to also be released, wait until all changes are in the public repo and let it soak there for some time. Then consider doing the backports for Indigo and Jade.

File: rostime/src/time.cpp (hunk 3 above)
Comment (flixr) [reply rostime/src/time.cpp:172]: I just copied that whole part from the implemtation of ros_walltime...

File: rostime/src/time.cpp (hunk 4 above)
Comment (flixr) [reply rostime/src/time.cpp:182]: Same as above: basically just copied that from ros_walltime...

File: rostime/src/time.cpp (hunk 5 above)
Comment (flixr) [reply rostime/src/time.cpp:188]: Same as above: basically just copied that from ros_walltime...

**Comment by flixr** [comment]
@dirk-thomas I would propose to merge this as is...
If you want further changes in the windows implementation, it should IMHO be done in a separate PR as the same should be applied to the already existing ros_walltime function and tested by someone on Windows...

**Comment by flixr** [comment]
@dirk-thomas any updates here?
I can't really test the windows code... but we have these patches running on our systems for a while now without any problems, so is there any reason to not merge this yet?

**Comment by flixr** [comment]
@dirk-thomas @mikaelarguedas ping...

**Comment by dirk-thomas** [comment]
Sorry for the delay - very busy times :worried:

I will merge this and release a new version into Lunar in order to test the ros_comm PR.

**Comment by dirk-thomas** [comment]
@flixr Thank you for working on this feature!
